packages/world/src/modules/keysintable/KeysInTableHook.sol

```diff
@@ -0,0 +1,80 @@
+import { KeysInTable } from "./tables/KeysInTable.sol";
+import { KeysInTableLength } from "./tables/KeysInTableLength.sol";
+import { UsedKeysIndex, UsedKeysIndexData } from "./tables/UsedKeysIndex.sol";
+import { ArrayLib } from "../utils/ArrayLib.sol";
+import { getTargetTableSelector } from "../utils/getTargetTableSelector.sol";
+
+/**
+ * This is a very naive and inefficient implementation for now.
+ * We can optimize this by adding support for `setIndexOfField` in Store
+ * and then replicate logic from solecs's Set.sol.
+ * (See https://github.com/latticexyz/mud/issues/444)
+ *
+ * Note: if a table with composite keys is used, only the first key is indexed
+ */
+contract KeysInTableHook is IStoreHook {
+  using ArrayLib for bytes32[];
+  using ResourceSelector for bytes32;
+
+  function handleSet(bytes32 tableId, bytes32[] memory key) internal {
+    bytes32 keysInTableTableId = getTargetTableSelector(MODULE_NAMESPACE, tableId);
+    bytes32 keysInTableLengthTableId = getTargetTableSelector(KEYS_LENGTH_NAMESPACE, tableId);
+    bytes32 usedIndexTableId = getTargetTableSelector(USED_KEYS_NAMESPACE, tableId);
+
+    bytes32 keysHash = keccak256(abi.encode(key));
+
+    // If the key not has yet been set in the table...
+    if (!UsedKeysIndex.getHas(usedIndexTableId, keysHash)) {
+      uint32 len = KeysInTableLength.get(keysInTableLengthTableId);
+
+      // Push the key to the list of keys in this table
+      KeysInTable.push(keysInTableTableId, key[0]);
+      KeysInTableLength.set(keysInTableLengthTableId, len + 1);
+
+      // Update the index to avoid duplicating this key in the array
+      UsedKeysIndex.set(usedIndexTableId, keysHash, UsedKeysIndexData(true, len));
+    }
+  }
+
+  function onSetRecord(bytes32 table, bytes32[] memory key, bytes memory) public {
+    handleSet(table, key);
+  }
+
+  function onBeforeSetField(bytes32 table, bytes32[] memory key, uint8, bytes memory) public {}
+
+  function onAfterSetField(bytes32 table, bytes32[] memory key, uint8, bytes memory) public {
+    handleSet(table, key);
+  }
+
+  function onDeleteRecord(bytes32 tableId, bytes32[] memory key) public {
+    bytes32 keysInTableTableId = getTargetTableSelector(MODULE_NAMESPACE, tableId);
+    bytes32 keysInTableLengthTableId = getTargetTableSelector(KEYS_LENGTH_NAMESPACE, tableId);
+    bytes32 usedIndexTableId = getTargetTableSelector(USED_KEYS_NAMESPACE, tableId);
+
+    bytes32 keysHash = keccak256(abi.encode(key));
+    UsedKeysIndexData memory data = UsedKeysIndex.get(usedIndexTableId, keysHash);
+
+    // If the key has not been set in the table...
+    if (data.has) {
+      uint32 len = KeysInTableLength.get(keysInTableLengthTableId);
+      bytes32 lastKey = KeysInTable.get(keysInTableTableId)[len - 1];
+
+      // Remove the key from the list of keys in this table
+      KeysInTable.update(keysInTableTableId, data.index, lastKey);
+      KeysInTable.pop(keysInTableTableId);
```

Comment:
You should be able to simplify this to:

```suggestion
      // Remove the key from the list of keys in this table
      bytes32 lastKey = KeysInTable.pop(keysInTableTableId);
      KeysInTable.update(keysInTableTableId, data.index, lastKey);
```

Reply:
I wanted to do that, but `pop()` doesn't currently return the key.